rio/src/RIO/File.hs

```diff
@@ -0,0 +1,345 @@
+-- bring benefits in terms of durability
+--
+-- @since 0.1.6
+ensureFileDurable :: MonadUnliftIO m => FilePath -> m ()
+ensureFileDurable absFp =
+#if WINDOWS
+  absFp `seq` return ()
+#else
+  bracket (openFileDurable absFp ReadMode)
+          (uncurry closeFileDurable)
+          (const $ return ())
+#endif
+
+
+-- | Similar to 'writeFileBinary', but it also ensures that changes executed to
+-- the file are guaranteed to be durable. It internally uses fsync and makes
+-- sure it synchronizes the file on disk.
+--
+-- @since 0.1.6
+writeFileBinaryDurable :: MonadUnliftIO m => FilePath -> ByteString -> m ()
+writeFileBinaryDurable absFp bytes =
+#if WINDOWS
+  writeFileBinary absFp bytes
+#else
+  withFileDurable absFp WriteMode (liftIO . (`hPut` bytes))
+#endif
+
+-- | Similar to 'writeFileBinary', but it also guarantes that changes executed
+-- to the file are durable, also, in case of failure, the modified file is never
+-- going to get corrupted. It internally uses fsync and makes sure it
+-- synchronizes the file on disk.
+--
+-- @since 0.1.6
+writeFileBinaryDurableAtomic :: MonadUnliftIO m => FilePath -> ByteString -> m ()
+writeFileBinaryDurableAtomic fp bytes =
+#if WINDOWS
+  writeFileBinary fp bytes
+#else
+  withFileDurableAtomic fp WriteMode (liftIO . (`hPut` bytes))
+#endif
+
+-- | Opens a file with the following guarantees:
+--
+-- * Successfully closes the file in case of an asynchronous exception
+--
+-- * It reliably saves the file in the correct directory; including edge case
+--   situations like a different device being mounted to the current directory,
+--   or the current directory being renamed to some other name while the file is
+--   being used.
+--
+-- * It ensures durability by executing fsync before close
+--
+-- @since 0.1.6
+withFileDurable ::
+     MonadUnliftIO m => FilePath -> IOMode -> (Handle -> m r) -> m r
+withFileDurable absFp iomode cb =
+#if WINDOWS
+  withFile absFp iomode cb
+#else
+  withRunInIO $ \run ->
+    bracket
+      (openFileDurable absFp iomode)
+      (uncurry closeFileDurable)
+      (\(_, h) -> do run (cb h))
+#endif
+
+-- | Opens a file with the following guarantees:
+--
+-- * Successfully closes the file in case of an asynchronous exception
+--
+-- * It reliably saves the file in the correct directory; including edge case
+--   situations like a different device being mounted to the current directory,
+--   or the current directory being renamed to some other name while the file is
+--   being used.
+--
+-- * It ensures durability by executing fsync before close
+--
+-- * It keeps all changes in a copy file, and after is closed it renames it to
+--   the original filepath, in case of catastrophic failure, the original file
+--   stays unaffected.
+--
+--
+-- === Performance Considerations
+--
+-- When using an 'IOMode' of 'WriteMode', 'ReadWriteMode', or 'AppendMode', this function
+-- performs a copy operation of the file being read. This may be prohibitive in
```

Comment:
I don't understand why we need to do a copy here instead of doing an atomic move.

Reply:
Well, my thoughts were, in case there is a hard restart in the middle of the write operation, the original file should stay the same, by moving the file, the guarantee is not there. This is a very expensive operation though

Reply:
I'm still not following, sorry. The typical approach for atomic file writing is to write the full contents to a temp file in the same directory and then rename the file atomically. I don't see where a file copy protects against a hard restart in the middle of a write operation.

Reply:
Ok, it is likely I'm missing something here, so you tell me how a move would work in the following example:

I have a `db.txt` file on my program that I want to open in `ReadWriteMode`, and I use a `withBinaryFileDurableAtomic` call, I do some reads on the file and then some other writes:

```haskell
withBinaryFileDurableAtomic "db.txt" ReadWriteMode $ \h ->
  -- some reads (1)
  -- some writes start (2)
  -- calculations that take considerate time (3)
  -- some writes (4)
```

Under the assumption that the system restarts/crashes in step (3), I understand that if we 

(a) do a copy of `db.txt` to  a temp file `.db.txt123abc.swp` and rename back after successful writes are done

both files, an unmodified `db.txt` and a temp file copy `.db.txt123abc.swp` files are going to be present in the file system. If we run the algorithm again, is going to start as if `db.txt` was never modified.

(b) do a move/rename of `db.txt` to `.db.txt123abc.swp` and rename back after successful writes are done

the original file `db.txt` is not going to be present at the restart, only the unmodified `.db.txt123abc.swp` file, which would make the original algorithm not work.

I'm now noticing you are mentioning an _atomic_ move, which maybe it means something different from a regular `mv`/`rename` call.

Apologies for not making a better effort at explaining my thoughts before and thank you for helping me figure this out and spend some minutes going through this.

Reply:
Atomic move is the same as `mv`/`rename`, sorry for the confusion.

I think the issue here is that I wasn't thinking of `ReadWriteMode`, but rather `WriteMode`. For `WriteMode`, I'm fairly certain the copy isn't necessary, right? It's really only a question for `ReadWriteMode`. I'm not even convinced there that we want to copy, as opposed to moving to a temp file location and potentially losing the file on a crash.

Reply:
> I think the issue here is that I wasn't thinking of ReadWriteMode, but rather WriteMode. For WriteMode, I'm fairly certain the copy isn't necessary, right?

I'm not sure, I would say it depends if the file is open _only_ as a `WriteMode` on your program lifecycle, in this situation, this statement makes sense. So in this case, you want something that is atomic but not durable (you don't care about the previous save).

> I'm not even convinced there that we want to copy, as opposed to moving to a temp file location and potentially losing the file on a crash.

Well, I argue losing the given file on a crash should be a decision made by the caller of this API, not by us. IMO, there is no atomicity if you can lose the file on a crash, also, there is already a function that offers durability without atomicity that the caller can use if they don't want to incur on the `copy` costs.

On a different note, I updated the documentation of the module with more information as a recommendation from @nh2